spec/user_interface_spec.rb

```diff
@@ -72,4 +72,61 @@
 
     expect(ui.run).to eq(1)
   end
+
+  describe 'ask for fields method' do
+    let(:input) { StringIO.new(contact_inputs(TEST_DETAILS)) }
+    let(:ui) { described_class.new(input, output) }
+
+    it 'asks user for all fields' do
+      ui.ask_for_fields
+
+      expect(output.string).to include(described_class::FIELDS_TO_PROMPTS.values.join("\n"))
+    end
+
+    it 'reads the input for all prompt' do
+      user_input = ui.ask_for_fields
+
+      expect(user_input).to eq(TEST_DETAILS)
+    end
+
+    it 'sends error if number is invalid' do
+      input = StringIO.new(contact_inputs(INVALID_INPUTS))
+      ui = described_class.new(input, output)
+
+      ui.ask_for_fields
+
+      expect(output.string).to include(described_class::ERROR_MESSAGE)
+    end
+
+    it 'sends error if email is invalid' do
+      input = StringIO.new(contact_inputs(INVALID_INPUTS))
+      ui = described_class.new(input, output)
+
+      ui.ask_for_fields
+
+      expect(output.string).to include(described_class::ERROR_MESSAGE)
+    end
+  end
+
+  TEST_DETAILS = {
+    name: 'Matt Damon',
+    address: 'Some address',
+    phone: '08796564231',
+    email: 'matt@damon.com',
+    notes: 'I think he has an Oscar'
+  }.freeze
+
+  INVALID_INPUTS = {
```

Comment:
Don't have constants in tests.

It's better to have a helper method that has the data of the happy path (i.e. all valid values) and then you can call it with wrong inputs in the test that need wrong inputs, to make it very explicit what are the bad inputs, like this:

```ruby
def test_details(alternative_values = {})
  {
    name: 'Matt Damon',
    address: 'Some address',
    # ...etc
  }.merge(alternative_values)
end
```
  
Then in tests you can send it invalid values in a very explicit way:

```ruby
invalid_values = {
  phone: `not_a_phone`
}

test_details(invalid_values)
```

Reply:
I'm having trouble getting this to work as I need a specific order

Reply:
I see what you mean, but what you have done now looks better